File: character.py
from util import prof_bonus
from feats import Vex, Feat, Topple
from target import Target
from weapons import Weapon
from events import HitArgs, AttackRollArgs, AttackArgs, MissArgs, WeaponRollArgs
from log import log
from typing import List
from collections import defaultdict
from util import spell_slots, highest_spell_slot, lowest_spell_slot
from spells import Spell, Spellcaster
from typing import Callable, Any


class Character:
    def init(
        self,
        level=None,
        stats=None,
        feats=None,
        base_feats: List[Feat] = None,
        feat_schedule=[4, 8, 12, 16, 19],
        default_feats=None,
        spellcaster: Spellcaster = Spellcaster.NONE,
        spell_mod: str = None,
    ):
        if default_feats is None:
            default_feats = [Vex(), Topple()]
        self.level = level
        self.prof = prof_bonus(level)
        self.str = stats[0]
        self.dex = stats[1]
        self.con = stats[2]
        self.int = stats[3]
        self.wis = stats[4]
        self.cha = stats[5]
        self.minions = []
        self.effects = set()
        self.spellcaster = spellcaster
        self.spell_mod = spell_mod
        self.concentration = None
        self.masteries = []
        self.feats = dict()
        self.feats_by_event = dict()
        for feat in default_feats:
            self.add_feat(feat)
        for feat in base_feats:
            self.add_feat(feat)
        for [target, feat] in zip(feat_schedule, feats):
            if level >= target:
                self.add_feat(feat)

    def add_feat(self, feat):
        feat.apply(self)
        self.feats[feat.name] = feat
        for event in feat.events():
            if event not in self.feats_by_event:
                self.feats_by_event[event] = []
            self.feats_by_event[event].append(feat)

    def has_feat(self, name: str):
        return name in self.feats

    def feat(self, name: str):
        return self.feats[name]

    def feats_for_event(self, event: str):
        if event not in self.feats_by_event:
            return []
        return self.feats_by_event[event]

    def has_feats_for_event(self, event: str):
        return event in self.feats_by_event

    def mod(self, stat: str):
        if stat == "none":
            return 0
        return (self.__getattribute__(stat) - 10) // 2

    def dc(self, stat: str):
        return self.mod(stat) + self.prof + 8

    def add_minion(self, minion):
        self.minions.append(minion)

    def remove_minion(self, minion):
        self.minions.remove(minion)

    def add_effect(self, effect: str):
        self.effects.add(effect)

    def remove_effect(self, effect: str):
        if effect in self.effects:
            self.effects.remove(effect)

    def has_effect(self, effect: str):
        return effect in self.effects

    def use_bonus(self, source: str):
        if not self.used_bonus:
            log.record(f"Bonus ({source})", 1)
            self.used_bonus = True
            return True
        return False

    # =============================
    #       LIFECYCLE EVENTS
    # =============================

    def begin_turn(self, target: Target):
        log.record("Turn", 1)
        self.actions = 1
        self.used_bonus = False
        for feat in self.feats_for_event("begin_turn"):
            feat.begin_turn(target)

    def end_turn(self, target: Target):
        for feat in self.feats_for_event("end_turn"):
            feat.end_turn(target)
        if not self.used_bonus:
            log.record(f"Bonus (None)", 1)
        log.output(lambda: "")

    def turn(self, target: Target):
        self.begin_turn(target)
        self.before_action(target)
        while self.actions > 0:
            self.action(target)
            self.actions -= 1
        self.after_action(target)
        self.end_turn(target)
        for minion in self.minions:
            minion.turn(target)

    def before_action(self, target: Target):
        for feat in self.feats_for_event("before_action"):
            feat.before_action(target)

    def action(self, target: Target):
        for feat in self.feats_for_event("action"):
            feat.action(target)

    def after_action(self, target: Target):
        for feat in self.feats_for_event("after_action"):
            feat.after_action(target)

    def short_rest(self):
        self.effects = set()
        for feat in self.feats_for_event("short_rest"):
            feat.short_rest()

    def long_rest(self):
        self.reset_spell_slots()
        self.short_rest()
        for feat in self.feats_for_event("long_rest"):
            feat.long_rest()

    def enemy_turn(self, target: Target):
        for feat in self.feats_for_event("enemy_turn"):
            feat.enemy_turn(target)

    # ============================
    #      WEAPON ATTACKS
    # ============================

    def add_masteries(self, masteries: List[str]):
        self.masteries.extend(masteries)

    def has_mastery(self, mastery: str) -> bool:
        return mastery in self.masteries

    def get_weapon_mod(self, weapon: Weapon):
        if weapon.override_mod is not None:
            return weapon.override_mod
        elif weapon.has_tag("ranged"):
            return "dex"
        elif (weapon.has_tag("finesse") or weapon.has_tag("thrown")) and (
            self.mod("dex") > self.mod("str")
        ):
            return "dex"
        else:
            return "str"

    def attack(
        self,
        target: Target,
        weapon: Weapon,
        tags: List[str] = [],
    ):
        args = AttackArgs(
            character=self,
            target=target,
            weapon=weapon,
            tags=tags,
            mod=self.get_weapon_mod(weapon),
        )
        log.record(f"Attack:{args.weapon.name}", 1)
        self.before_attack()
        if weapon.override_to_hit:
            to_hit = weapon.override_to_hit()
        else:
            to_hit = self.prof + self.mod(args.mod) + weapon.attack_bonus
        result = self.roll_attack(attack=args, to_hit=to_hit)
        roll = result.roll()
        crit = False
        if roll >= args.weapon.min_crit:
            crit = True
        roll_total = roll + to_hit + result.situational_bonus
        log.output(lambda: f"{args.weapon.name} total {roll_total} vs {args.target.ac}")
        if roll_total >= args.target.ac:
            self.hit(attack=args, crit=crit, roll=roll)
        else:
            self.miss(attack=args)

    def before_attack(self):
        for feat in self.feats_for_event("before_attack"):
            feat.before_attack()

    def roll_attack(self, attack: AttackArgs, to_hit: int) -> AttackRollArgs:
        args = AttackRollArgs(attack=attack, to_hit=to_hit)
        if attack.target.stunned:
            args.adv = True
        if attack.target.prone:
            if args.attack.weapon.has_tag("ranged"):
                args.disadv = True
            else:
                args.adv = True
        if attack.target.semistunned:
            args.adv = True
            args.attack.target.semistunned = False
        for feat in self.feats_for_event("roll_attack"):
            feat.roll_attack(args)
        return args

    def weapon_roll(self, weapon: Weapon, crit: bool = False):
        rolls = weapon.rolls(crit=crit)
        args = WeaponRollArgs(weapon=weapon, rolls=rolls, crit=crit)
        for feat in self.feats_for_event("weapon_roll"):
            feat.weapon_roll(args)
        return sum(args.rolls)

    def hit(
        self,
        attack: AttackArgs,
        crit: bool = False,
        roll: int = 0,
    ):
        args = HitArgs(attack=attack, crit=crit, roll=roll)
        target = attack.target
        weapon = attack.weapon
        log.record(f"Hit:{weapon.name}", 1)
        if crit:
            log.record(f"Crit:{weapon.name}", 1)
        dmg = self.weapon_roll(weapon, crit=crit)
        if weapon.flat_dmg_bonus:
            dmg += weapon.flat_dmg_bonus
        else:
            dmg += weapon.dmg_bonus
            if not attack.has_tag("light"):
                mod = self.get_weapon_mod(weapon)
                dmg += self.mod(mod)
        args.add_damage(f"Weapon:{weapon.name}", dmg)
        for feat in self.feats_for_event("hit"):
            feat.hit(args)
        log.output(lambda: str(args._dmg))
        for key in args._dmg:
            target.damage_source(key, args.dmg_multiplier * args._dmg[key])

    def miss(self, attack: AttackArgs):
        args = MissArgs(attack)
        for feat in self.feats_for_event("miss"):
            feat.miss(args)

    # ==================================
    #         SPELLCASTING
    # ==================================

    def reset_spell_slots(self):
        self.slots = spell_slots(self.level, half=self.spellcaster is Spellcaster.HALF)

    def spell_dc(self):
        return 8 + self.mod(self.spell_mod) + self.prof

    def highest_slot(self, max: int = 9) -> int:
        return highest_spell_slot(self.slots, max=max)

    def lowest_slot(self, min: int = 1) -> int:
        return lowest_spell_slot(self.slots, min=min)

    def cast(self, spell: Spell, target: Target = None):
        if spell.slot > 0:
            self.slots[spell.slot] -= 1
        if spell.concentration:
            self.set_concentration(spell)
        spell.cast(self, target)

    def set_concentration(self, spell: Spell):
        if self.concentration:
            self.concentration.end(self)
        self.concentration = spell

    def concentrating_on(self, name: str) -> bool:
        return self.concentration is not None and self.concentration.name is name

    def is_concentrating(self) -> bool:
        return self.concentration is not None

File: dndsim/src/sim/Character.ts
import {
    CharacterEventMapping,
    CharacterEventName,
} from "./events/CharacterEvent"
import { Feat } from "./Feat"
import { EventLoop } from "../util/EventLoop"
import { Class, Stat, WeaponMastery } from "./types"
import { Resource } from "./resources/Resource"
import { Target } from "./Target"
import { Weapon } from "./Weapon"
import { Attack, WeaponAttack } from "./Attack"
import { DamageRoll } from "./helpers/DamageRoll"
import { Spellcasting } from "./spells/Spellcasting"
import { Spell } from "./spells/Spell"
import { ActionEvent } from "./events/ActionEvent"
import { AttackEvent } from "./events/AttackEvent"
import { AttackResultEvent } from "./events/AttackResultEvent"
import { AttackRollEvent } from "./events/AttackRollEvent"
import { DamageRollEvent } from "./events/DamageRollEvent"
import { BeforeAttackEvent } from "./events/BeforeAttackEvent"
import { BeforeActionEvent } from "./events/BeforeActionEvent"
import { AfterActionEvent } from "./events/AfterActionEvent"
import { EndTurnEvent } from "./events/EndTurnEvent"
import { BeginTurnEvent } from "./events/BeginTurnEvent"
import { ShortRestEvent } from "./events/ShortRestEvent"
import { LongRestEvent } from "./events/LongRestEvent"
import { EnemyTurnEvent } from "./events/EnemyTurnEvent"
import { NumAttacksAttribute } from "./actions/AttackAction"
import { Graze } from "./coreFeats/Graze"
import { Topple } from "./coreFeats/Topple"
import { Vex } from "./coreFeats/Vex"
import { log } from "../util/Log"
import { CombatSuperiority } from "./resources/CombatSuperiority"
import { CustomTurn } from "./actions/CustomTurn"
import { Environment } from "./Environment"

const DEFAULT_STAT_MAX = 20

export class Character {
    // Stats
    level: number = 0
    stats: Record<Stat, number>
    statMax: Record<Stat, number> = {
        str: DEFAULT_STAT_MAX,
        dex: DEFAULT_STAT_MAX,
        con: DEFAULT_STAT_MAX,
        int: DEFAULT_STAT_MAX,
        wis: DEFAULT_STAT_MAX,
        cha: DEFAULT_STAT_MAX,
        none: DEFAULT_STAT_MAX,
    }

    feats: Array<Feat> = []
    events: EventLoop<CharacterEventName, CharacterEventMapping> =
        new EventLoop()
    effects: Set<string> = new Set()
    minions: Set<Character> = new Set()

    spells: Spellcasting = new Spellcasting(this)
    masteries: Set<WeaponMastery> = new Set()
    classLevels: Map<Class, number> = new Map()
    attributes: Map<string, number> = new Map([[NumAttacksAttribute, 1]])

    // Resources
    bonus = new Resource({
        name: "Bonus",
        character: this,
        initialMax: 1,
    })
    actions = new Resource({
        name: "Action",
        character: this,
        initialMax: 1,
    })
    combatSuperiority: CombatSuperiority = new CombatSuperiority(this)
    heroicInspiration = new Resource({
        name: "HeroicInspiration",
        character: this,
        initialMax: 1,
    })
    ki = new Resource({
        name: "Ki",
        character: this,
        resetOnLongRest: true,
    })
    resources: Map<string, Resource> = new Map()
    // TODO: Add other class resources
    // TODO: Handle actions better
    grappleStat: Stat = "str"
    customTurn: CustomTurn = new CustomTurn()

    constructor(args: { stats: Omit<Record<Stat, number>, "none"> }) {
        const { stats } = args
        this.stats = { ...stats, none: 10 }
        for (const feat of [new Graze(), new Topple(), new Vex()]) {
            this.addFeat(feat)
        }
    }

    addFeat(feat: Feat): void {
        this.feats.push(feat)
        feat.internalApply(this)
    }

    stat(stat: Stat): number {
        return this.stats[stat]
    }

    mod(stat: Stat): number {
        return Math.floor((this.stat(stat) - 10) / 2)
    }

    increaseStatMax(stat: Stat, amount: number): void {
        this.statMax[stat] += amount
    }

    increaseStat(stat: Stat, amount: number): void {
        this.stats[stat] += amount
        if (this.stats[stat] > this.statMax[stat]) {
            this.stats[stat] = this.statMax[stat]
        }
    }

    increaseStatAndMax(stat: Stat, amount: number) {
        this.increaseStatMax(stat, amount)
        this.increaseStat(stat, amount)
    }

    dc(stat: Stat): number {
        return this.mod(stat) + this.prof() + 8
    }

    prof(): number {
        return Math.floor((this.level - 1) / 4) + 2
    }

    addMinion(minion: Character): void {
        this.minions.add(minion)
    }

    removeMinion(minion: Character): void {
        this.minions.delete(minion)
    }

    hasEffect(effect: string): boolean {
        return this.effects.has(effect)
    }

    addEffect(effect: string): void {
        this.effects.add(effect)
    }

    removeEffect(effect: string): void {
        this.effects.delete(effect)
    }

    getAttribute(attribute: string): number {
        return this.attributes.get(attribute) ?? 0
    }

    setAttribute(attribute: string, value: number): void {
        this.attributes.set(attribute, value)
    }

    getClassLevel(class_: Class): number {
        return this.classLevels.get(class_) ?? 0
    }

    hasClassLevel(class_: Class, level: number): boolean {
        return (this.getClassLevel(class_) ?? 0) >= level
    }

    addClassLevel(class_: Class, level: number): void {
        this.classLevels.set(class_, this.getClassLevel(class_) + level)
        this.level += level
    }

    getResource(name: string): Resource {
        if (!this.resources.has(name)) {
            this.resources.set(
                name,
                new Resource({
                    name,
                    character: this,
                    initialMax: 0,
                })
            )
        }
        return this.resources.get(name)!
    }

    hasResource(name: string): boolean {
        return this.resources.get(name)?.has() ?? false;
    }

    useResource(name: string) {
        this.getResource(name).use()
    }

    // =============================
    //       COMMON ACTIONS
    // =============================

    grapple(target: Target): void {
        if (target.save(this.dc(this.grappleStat))) {
            target.grapple()
        }
    }

    // =============================
    //       LIFECYCLE EVENTS
    // =============================

    turn(target: Target): void {
        log.record("Turn", 1)
        this.actions.reset()
        this.bonus.reset()
        if (this.customTurn.hasOperations()) {
            this.customTurn.doTurn(
                new Environment({ character: this, target }),
                this
            )
        } else {
            this.events.emit("begin_turn", new BeginTurnEvent({ target }))
            this.events.emit("before_action", new BeforeActionEvent({ target }))
            while (this.actions.use()) {
                this.events.emit("action", new ActionEvent({ target }))
            }
            this.events.emit("after_action", new AfterActionEvent({ target }))
            this.events.emit("end_turn", new EndTurnEvent({ target }))
        }
        for (const minion of this.minions) {
            minion.turn(target)
        }
    }

    shortRest(): void {
        this.effects = new Set()
        this.events.emit("short_rest", new ShortRestEvent())
    }

    longRest(): void {
        this.shortRest()
        this.events.emit("long_rest", new LongRestEvent())
    }

    enemyTurn(target: Target): void {
        this.events.emit("enemy_turn", new EnemyTurnEvent({ target }))
    }

    // =============================
    //       WEAPON EVENTS
    // =============================

    weaponAttack(args: {
        target: Target
        weapon: Weapon
        tags?: string[]
    }): void {
        const attack = new WeaponAttack({
            weapon: args.weapon,
            tags: args.tags,
        })
        this.attack({ target: args.target, attack })
    }

    attack(args: { target: Target; attack: Attack }): void {
        const { target, attack } = args
        log.record(`Attack (${attack.name()})`, 1)
        const attackData = new AttackEvent({ target, attack })
        this.events.emit("before_attack", new BeforeAttackEvent())
        const toHit = attack.toHit(this)
        const rollResult = this.attackRoll(attackData, toHit)
        const roll = rollResult.roll()
        const minCrit = rollResult.minCrit
            ? rollResult.minCrit
            : attack.minCrit()
        const crit = roll >= minCrit
        const rollTotal = roll + toHit + rollResult.situationalBonus
        const hit = rollTotal >= target.ac

        if (hit) {
            log.record(`Hit (${attack.name()})`, 1)
        } else {
            log.record(`Miss (${attack.name()})`, 1)
        }
        if (crit) {
            log.record(`Crit (${attack.name()})`, 1)
        }

        const attackResult = new AttackResultEvent({
            attack: attackData,
            hit,
            crit,
            roll,
        })
        args.attack.attackResult(attackResult, this)
        this.events.emit("attack_result", attackResult)
        for (const damage of attackResult.damageRolls) {
            if (crit) {
                damage.dice = damage.dice.concat(damage.dice)
            }
            this.doDamage({
                target,
                damage,
                attack: attackData,
                multiplier: attackResult.dmgMultiplier,
            })
        }
    }

    attackRoll(attackData: AttackEvent, toHit: number): AttackRollEvent {
        const { target, attack } = attackData
        const data = new AttackRollEvent({ attack: attackData, toHit })
        // TODO: Stunned and semistunned
        if (target.prone) {
            if (attack.isRanged()) {
                data.disadv = true
            } else {
                data.adv = true
            }
        }
        this.events.emit("attack_roll", data)
        return data
    }

    doDamage(args: {
        target: Target
        damage: DamageRoll
        attack?: AttackEvent
        multiplier?: number
        spell?: Spell
    }): void {
        const { target, damage, attack, spell } = args
        const multiplier = args.multiplier ?? 1
        const damageData = new DamageRollEvent({
            target,
            damage,
            attack,
            spell,
        })
        this.events.emit("damage_roll", damageData)
        target.addDamage(damage.source, Math.floor(damage.total() * multiplier))
    }
}
